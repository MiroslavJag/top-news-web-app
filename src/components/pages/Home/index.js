import React from 'react'
import styled from 'styled-components'

const Content = styled.h1`
  margin: 0;
`

const Home = () => {
  return (
    <>
      <Content>I am Home</Content>
      <Content>I am Home</Content>
      <Content>I am Home</Content>
      <Content>I am Home</Content>
      <Content>I am Home</Content>
      <Content>I am Home</Content>
      <Content>I am Home</Content>
      <Content>I am Home</Content>
      <Content>I am Home</Content>
      <Content>I am Home</Content>
      <Content>I am Home</Content>
      <Content>I am Home</Content>
      <Content>I am Home</Content>
      <Content>I am Home</Content>
      <Content>I am Home</Content>
      <Content>I am Home</Content>
      <Content>I am Home</Content>
      <Content>I am Home</Content>
      <Content>I am Home</Content>
      <Content>I am Home</Content>
      <Content>I am Home</Content>
      <Content>I am Home</Content>
      <Content>I am Home</Content>
      <Content>I am Home</Content>
      <Content>I am Home</Content>
      <Content>I am Home</Content>
      <Content>I am Home</Content>
      <Content>I am Home</Content>
    </>
  )
}

export default Home
