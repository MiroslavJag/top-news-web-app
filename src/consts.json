{
 "ERRORS": {
  "API_ERROR": "API ERROR"
 },
 "LANG_CODES": {
  "GB": "gb",
  "US": "us"
 },
 "TOP_NEWS_LANG": "TopNewsLang",
 "CATEGORY_NAME": {
  "BUSINESS": "business",
  "ENTERTAINMENT": "entertainment",
  "GENERAL": "general",
  "HEALTH": "health",
  "SCIENCE": "science",
  "SPORTS": "sports",
  "TECHNOLOGY": "technology"
 }
}