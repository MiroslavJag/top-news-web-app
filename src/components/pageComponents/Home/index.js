import React, {useEffect, useState} from 'react'
import styled from 'styled-components'
import {Link} from 'react-router-dom'
import {connect} from 'react-redux'
import * as actions from 'store/actions'
import {getTopNews} from '../../../services/topNews'
import ErrorView from 'components/Errors/ErrorView'
import {reducePlaceDescLength} from '../../../helpers/strings'
import NoImage from 'static/images/no-image-thumbnail.jpg'
import moment from 'moment'
import {useTranslation} from 'react-i18next'
import SingleNews from 'components/common/SingleNews'
import SkeletonLoader from 'components/common/SkeletonLoader'
import NewsList from 'components/common/NewsList'
import {LANG_CODES} from 'consts.json'
import {API_ROUTES} from 'routes.json'


export const NewsListTitleWrapper = styled.div`
  display: flex;
  flex-direction: row;
  border-bottom: 1px solid lightgrey;
  margin: 0 0 20px 0;
`

export const NewsListTitle = styled.h1`
  margin: 0 0 10px 0;
  text-align: left;
`
export const NewsListTitleLang = styled(NewsListTitle)`
  text-transform: uppercase;
`
export const Separator = styled.div`
  margin: 0px 15px 0;
  height: 35px;
  width: 1px;
  background-color: lightgrey;
`

const NewsSectionWrapper = styled.div`
  width: 100%;
  justify-content: center;
`

export const NewsPublishedTime = styled.p`
  margin: 0 0 10px 0;
  font-size: 13px;
  color: grey;
`

const LoadingSkeltonWrapper = styled.div`
  background-color: lightgrey;
  border-radius: 4px;
  position: relative;
  overflow: hidden;
  height: 100px;
  width: 100%;

  &::before {
    content: '';
    display: block;
    position: absolute;
    left: -150px;
    top: 0;
    height: 100%;
    width: 150px;
    background: linear-gradient(to right, transparent 0%, #e8e8e8 50%, transparent 100%);
    animation: load 2s cubic-bezier(0.4, 0, 0.2, 1) infinite;
  }

  @keyframes load {
    from {
      left: -150px;
    }
    to {
      left: 100%;
    }
  }
`

type Props = {
  langCode: String,
  setLanguageDisabled: Function,
}

const Home = ({langCode, setLanguageDisabled}): Props => {
  const [data, setData] = useState(null)
  const [singleNewsData, setSingleNewsData] = useState(null)
  const {t} = useTranslation()

  useEffect(() => {
    fetchTopNews()
  }, [langCode])

  useEffect(() => {
    if (singleNewsData !== null) {
      setLanguageDisabled(true)
    }
  }, [singleNewsData])

  const fetchTopNews = async () => {
    const newsData = await getTopNews(API_ROUTES.TOP_HEADLINES, langCode)
    setData(newsData)
  }


  const renderTopNewsContent = () =>
    data !== null ? (
      data.error ? (
        <ErrorView refreshCallback={fetchTopNews} />
      ) : (
        <>
          <NewsListTitleWrapper>
            <NewsListTitle>{t('TOP_NEWS_TITLE')}</NewsListTitle>
            <Separator />
            <NewsListTitleLang>{langCode}</NewsListTitleLang>
          </NewsListTitleWrapper>
          <NewsList data={data.articles} setSingleNewsData={setSingleNewsData} />
        </>
      )
    ) : (
      <SkeletonLoader langCode={langCode} />
    )

  return (
    <>
      {singleNewsData !== null ? (
        <SingleNews singleNewsData={singleNewsData} setSingleNewsData={setSingleNewsData} />
      ) : (
        renderTopNewsContent()
      )}
    </>
  )
}
const mapStateToProps = (state) => {
  return {
    langCode: state.langCode,
  }
}
const mapDispatchToProps = (dispatch) => {
  return {
    setLanguageDisabled: (value) => dispatch(actions.setLanguageDisabled(value)),
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(Home)
