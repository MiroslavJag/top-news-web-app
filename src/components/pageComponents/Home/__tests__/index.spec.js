import React from 'react'
import {mount} from 'enzyme'
import renderer from 'react-test-renderer'
import 'jest-styled-components'
import Home from 'components/pageComponents/Home'
import {BrowserRouter} from 'react-router-dom'
import {Provider} from 'react-redux'
import configureStore from 'redux-mock-store'
import {LANG_CODES} from 'consts.json'
import {topNewsStub} from '../../../../test/stubs/news'

const mockStore = configureStore([])

describe('<Home />', () => {
  let store
  let component

  beforeEach(() => {
    store = mockStore({
      langCode: LANG_CODES.GB,
      langDisabled: false,
    })

    component = (
      <Provider store={store}>
        <BrowserRouter>
          <Home data={topNewsStub.articles} t={(k) => k} />
        </BrowserRouter>
      </Provider>
    )
  })

  it('match snapshot', () => {
    const tree = renderer.create(component).toJSON()
    expect(tree).toMatchSnapshot()
  })
  it('should render correctly with all needed components', () => {
    const wrapper = mount(component)
    expect(wrapper.find('Home__NewsListTitleWrapper')).toHaveLength(1)
    expect(wrapper.find('NewsList__NewsListWrapper')).toHaveLength(1)
  })
})
