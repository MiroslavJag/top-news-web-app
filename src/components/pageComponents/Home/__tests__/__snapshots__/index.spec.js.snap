// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> match snapshot 1`] = `
Array [
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  border-bottom: 1px solid lightgrey;
  margin: 0 0 20px 0;
}

.c1 {
  margin: 0 0 10px 0;
  text-align: left;
}

.c3 {
  text-transform: uppercase;
}

.c2 {
  margin: 0px 15px 0;
  height: 35px;
  width: 1px;
  background-color: lightgrey;
}

@media only screen and (min-width:768px) {

}

@media only screen and (min-width:1024px) {

}

<div
    className="c0"
  >
    <h1
      className="c1"
    >
      TOP_NEWS_TITLE
    </h1>
    <div
      className="c2"
    />
    <h1
      className="c1 c3"
    >
      gb
    </h1>
  </div>,
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c1 {
  max-width: 500px;
  width: 100%;
}

.c2 {
  background-color: lightgrey;
  position: relative;
  overflow: hidden;
  height: 100px;
  width: 100%;
}

.c2::before {
  content: '';
  display: block;
  position: absolute;
  left: -150px;
  top: 0;
  height: 100%;
  width: 150px;
  background: linear-gradient(to right,transparent 0%,#e8e8e8 50%,transparent 100%);
  -webkit-animation: load 2s cubic-bezier(0.4,0,0.2,1) infinite;
  animation: load 2s cubic-bezier(0.4,0,0.2,1) infinite;
}

.c3 {
  height: 20px;
  margin-bottom: 10px;
}

.c4 {
  height: 200px;
  margin-bottom: 10px;
}

.c5 {
  height: 40px;
  margin-bottom: 40px;
}

@media only screen and (min-width:768px) {
  .c1 {
    max-width: 300px;
    margin: 20px 5px 20px 20px;
  }
}

@media only screen and (min-width:1024px) {
  .c1 {
    max-width: 350px;
  }
}

<div
    className="c0"
  >
    <div
      className="c1"
    >
      <div
        className="c2 c3"
      />
      <div
        className="c2 c3"
      />
      <div
        className="c2 c3"
        style={
          Object {
            "width": "100px",
          }
        }
      />
      <div
        className="c2 c4"
      />
      <div
        className="c2 c5"
      />
    </div>
    <div
      className="c1"
    >
      <div
        className="c2 c3"
      />
      <div
        className="c2 c3"
      />
      <div
        className="c2 c3"
        style={
          Object {
            "width": "100px",
          }
        }
      />
      <div
        className="c2 c4"
      />
      <div
        className="c2 c5"
      />
    </div>
    <div
      className="c1"
    >
      <div
        className="c2 c3"
      />
      <div
        className="c2 c3"
      />
      <div
        className="c2 c3"
        style={
          Object {
            "width": "100px",
          }
        }
      />
      <div
        className="c2 c4"
      />
      <div
        className="c2 c5"
      />
    </div>
    <div
      className="c1"
    >
      <div
        className="c2 c3"
      />
      <div
        className="c2 c3"
      />
      <div
        className="c2 c3"
        style={
          Object {
            "width": "100px",
          }
        }
      />
      <div
        className="c2 c4"
      />
      <div
        className="c2 c5"
      />
    </div>
    <div
      className="c1"
    >
      <div
        className="c2 c3"
      />
      <div
        className="c2 c3"
      />
      <div
        className="c2 c3"
        style={
          Object {
            "width": "100px",
          }
        }
      />
      <div
        className="c2 c4"
      />
      <div
        className="c2 c5"
      />
    </div>
    <div
      className="c1"
    >
      <div
        className="c2 c3"
      />
      <div
        className="c2 c3"
      />
      <div
        className="c2 c3"
        style={
          Object {
            "width": "100px",
          }
        }
      />
      <div
        className="c2 c4"
      />
      <div
        className="c2 c5"
      />
    </div>
  </div>,
]
`;
